okrs_2019.md

```diff
@@ -0,0 +1,209 @@
+# SourceCred 2019 OKRs
+
+This document contains SourceCred's objectives, initiatives, and metrics for
+2019, generated as part of Protocol Lab's annual planning process.
+For an overview of SourceCred itself, see [the overview](overview.md).
+
+The purpose of this document is to provide clarity around what the project's
+goals and priorities are, and accountability to see if we are making effective
+progress towards achieving those goals. However, this is not intended to lock
+us in to a particular set of initiatives--we know that as time goes on, new
+opportunities will present themselves, and some of the existing initiatives
+will prove ill-founded or infeasible. Also, our metrics have known deficiencies
+(e.g. we don't yet know how to measure the quality of a cred attribution).
+
+As such, this is intended to be a living document. We will update it both to
+record progress on our goals, and to modify and retarget our goals as we gain
+more information. Since this document is in Git, all revisions will be
+publically visible. If we remove or replace a goal, initiave, or metric, we'll
+clearly explain the rationale for doing so in the associated commit.
+
+
+# SourceCred Goals
+
+## Improve Attribution Quality
+
+SourceCred's core function is to produce cred attributions for open-source
+projects. Improving the quality of those attributions is the our highest
+priority. We want SourceCred to do a great job of assigning cred both to
+contributors, and also to individual contributions.
+
+Assessing the value of code and code contributions will be very important, so a
+key initiative for 2019 is to add code analysis to SourceCred. However, coding
+is just one of the essential kinds of contributions that sustain an open-source
+project. We want to make demonstrable progress in recognizing some other
+important forms of project stewardship, such as:
+
+#### Project Stewardship Tasks
+- providing mentorship within the community
+- transferring context on the project, its history and design
+- writing documentation
+- improving test coverage
+- maintaining the build
+- triaging issues
+- helping to resolve interpersonal conflicts
+- planning for future development
+- refactoring difficult code
+- paying down technical debt
+- making the project fun to work on
+
+### Initiatives
```

Comment:
I want to call out here that a number of these can be done, to put it
bluntly, either “hastily” or “thoughtfully”. To give a few examples
(most of which we’ve talked about offline to some degree):

  - In terms of code analysis: What does it mean for “files” to be
    tracked by SourceCred? What is a “file”, given the context of a
    commit history? How are (say) functions identified across files and
    across time? What happens when someone refactors a function? What
    happens at merge commits? How can we make our analyses
    computationally tractable (do we need something like Hercules?)?

  - In terms of time-based cred analysis: What exactly do time-weights
    apply to (nodes? edges? something else?)? How does time-weighting
    interact with the inherent transitivity of the graph? If an
    unappreciated contribution from 2010 is made much more valuable by a
    contribution from 2015, how does the score for that 2010
    contribution change as a function of the weighting? For what
    purposes is time-weighted cred intended to be used?

  - In terms of heuristics: How do heuristics compose? When multiple
    heuristics apply to an item, with either contradictory or
    reinforcing assessments, how do we compile those evaluations? How do
    heuristics interact with “pass-through” semantics (a judgment that
    “this code reformatting commit is unimportant” may indicate to
    reduce the commit’s overall influence on the graph, whereas a
    judgment that “this merge commit is unimportant” more indicates that
    the merge commit’s incoming cred is still valuable but should be
    funneled through to its parents rather than its authors)?

  - In terms of “cred decomposition”: Should this satisfy some kind of
    additive properties, like “the cred to Node X is equal to the sum of
    (cred to Node X due to nodes in Part P) across Parts P that
    partition the space? Is there some natural interpretation of
    weighted cred decomposition? What does it mean to
    isolate/scope/decompose cred in such a cyclic structure?

Each of the above should have a design doc that specifies what
objectives were considered, what objectives were out of scope, what
implementations were considered, the relative advantages and
disadvantages of the implementations, and why the final decision was so
made. These are critical parts of SourceCred; you can’t just do them on
the fly.

Reply:
Agreed on all counts. I re-wrote this section of the document to try to incorporate your feedback, ptal.